### PR TITLE
Add parameter support to control stop or terminate an EC2 instance on shutdown in the ec2 module

### PR DESCRIPTION
**Pull request by SonOfBytes**
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

cloud/amazon/ec2.py
##### ANSIBLE VERSION

ansible 2.0.1.0
##### SUMMARY

Add support to the existing amazon/ec2 module to allow the parameter configuration of the shutdown behavior when creating an EC2 instance.

When creating an EC2 instance through the console it provides the option to Stop or Terminate an instance when the OS shuts down.  This change provides the same functionality from the module.

This is based on the abandoned PR https://github.com/ansible/ansible-modules-core/pull/1914

Supported by the issue raised here https://github.com/ansible/ansible-modules-core/issues/475

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @SonOfBytes for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2.py [hunk 1]
```diff
@@ -205,6 +205,13 @@
     required: false
     default: no
     choices: [ "yes", "no" ]
+  instance_initiated_shutdown_behavior:
+    version_added: "2.1"
+    description:
+    - Set whether AWS will Stop or Terminate an instance on shutdown
+    required: false
+    default: 'stop'
+    choices: [ "stop", "terminate"
```
Comment (ryansb) [inline cloud/amazon/ec2.py]: Close this brace please :smile:

File: cloud/amazon/ec2.py [hunk 2]
```diff
@@ -1033,6 +1041,9 @@ def create_instances(module, ec2, vpc, override_count=None):
                       private_ip_address = private_ip,
                     ))
 
+                # Spot instances do not support start/stop thereby not having the option to change shutdown behavior
+                params['instance_initiated_shutdown_behavior'] = instance_initiated_shutdown_behavior
```
Comment (ryansb) [inline cloud/amazon/ec2.py]: Wouldn't these 3 lines be unnecessary? The comment would still be useful on line 1081.

File: cloud/amazon/ec2.py (hunk 1 above)
Comment (SonOfBytes) [inline cloud/amazon/ec2.py]: will do

**Comment by SonOfBytes** [comment]
@ryansb braces fixed - thanks for pointing that out!

**Comment by gregdek** [comment]
Thanks @SonOfBytes for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by SonOfBytes** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @SonOfBytes for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @SonOfBytes for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@SonOfBytes A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@SonOfBytes Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@SonOfBytes A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by radamsfc** [comment]
ok continuing the game of whack-a-mole - looking at this again :)

**Comment by gregdek** [comment]
@SonOfBytes Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by SonOfBytes** [comment]
Confirmed
- No Merge Conflicts
- shippable passed

But for some unknown reason this is failing on some patch that is unrelated to the actual file being updated and this used to work.
- continuous-integration/travis-ci/pr failed

```
Applying: Add supports_check_mode=True to mysql_db module
Using index info to reconstruct a base tree...
M   database/mysql/mysql_db.py
Falling back to patching base and 3-way merge...
Auto-merging database/mysql/mysql_db.py
CONFLICT (content): Merge conflict in database/mysql/mysql_db.py
Failed to merge in the changes.
Patch failed at 0003 Add supports_check_mode=True to mysql_db module
The copy of the patch that failed is found in:
   /home/travis/build/ansible/ansible-modules-core/.git/rebase-apply/patch
```

Please advise if this needs further attention and how to address the issue if need be.

ready_for_review

**Comment by gregdek** [comment]
Thanks @SonOfBytes for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ryansb** [comment]
Works great - I rebased and committed as https://github.com/ansible/ansible-modules-core/commit/c8ca1a62118c0ffc9f3f4b1b3872764c141a9617. Thanks for your contribution!

**Pull request closed without merging**